numba/np/arraymath.py

```diff
@@ -864,6 +864,56 @@ def flat_all(a):
     return flat_all
 
 
+@overload(np.allclose)
+@overload_method(types.Array, "allclose")
+def np_allclose(a, b, rtol=1e-05, atol=1e-08, equal_nan=False):
+
+    if not type_can_asarray(a):
+        raise TypeError('The first argument "a" must be array-like')
+
+    if not type_can_asarray(b):
+        raise TypeError('The first argument "b" must be array-like')
+
+    if not isinstance(rtol, types.Float):
+        raise TypeError('The fourth argument "rtol" must be a '
+                        'floating point')
+
+    if not isinstance(atol, types.Float):
+        raise TypingError('The fifth argument "atol" must be a '
+                          'floating point')
+
+    if not isinstance(equal_nan, types.Boolean):
+        raise TypeError('The sixth argument "equal_nan" must be a '
+                        'boolean')
+
+    def np_allclose_impl(a, b, rtol=1e-05, atol=1e-08, equal_nan=False):
+        for av, bv in zip(np.nditer(a), np.nditer(b)):
```

Comment:
May be we need to broadcast `a` or `b` if needed. I am yet to do that.

Reply:
This doesn't work when `a` and `b` have different ranks because rank is a part of array type which gives an error like, `Failed to unify Array(float64, 2d, c) and Array(float64, 0d, c)`.  @sklam Any ideas for a workaround for the above problem?

Reply:
Do you have an input that replicate this error?

Reply:
**Code**

```python
from numba import jit
import numpy as np

def np_allclose(a, b, rtol=1e-05, atol=1e-08, equal_nan=False):
    return np.allclose(a, b, rtol, atol, equal_nan)

def test_allclose_1():

    pyfunc = np_allclose
    cfunc = jit(nopython=True)(pyfunc)

    zero_array = np.zeros((25, 4))
    b = np.asarray([1e-6])
    print("Shapes: ", zero_array.shape, b.shape)
    print(cfunc(zero_array, b))

test_allclose_1()
```

**Error**

```python
Shapes:  (25, 4) (1,)
Traceback (most recent call last):
  File "../debug.py", line 75, in <module>
    test_allclose_1()
  File "../debug.py", line 73, in test_allclose_1
    print(cfunc(zero_array, b))
  File "/Users/czgdp1807/numba_project/numba/numba/core/dispatcher.py", line 468, in _compile_for_args
    error_rewrite(e, 'typing')
  File "/Users/czgdp1807/numba_project/numba/numba/core/dispatcher.py", line 409, in error_rewrite
    raise e.with_traceback(None)
numba.core.errors.TypingError: Failed in nopython mode pipeline (step: nopython frontend)
No implementation of function Function(<function allclose at 0x7f7e9072f560>) found for signature:
 
 >>> allclose(array(float64, 2d, C), array(float64, 1d, C), float64, float64, Literal[bool](False))
 
There are 2 candidate implementations:
  - Of which 2 did not match due to:
  Overload in function 'np_allclose': File: numba/np/arraymath.py: Line 898.
    With argument(s): '(array(float64, 2d, C), array(float64, 1d, C), float64, float64, bool)':
   Rejected as the implementation raised a specific error:
     TypingError: Failed in nopython mode pipeline (step: nopython frontend)
   Cannot unify array(float64, 1d, C) and readonly array(float64, 2d, A) for 'b.1.2', defined at /Users/czgdp1807/numba_project/numba/numba/np/arraymath.py (932)
   
   File "numba/np/arraymath.py", line 932:
       def np_allclose_impl(a, b, rtol=1e-05, atol=1e-08, equal_nan=False):
           <source elided>
   
           for av, bv in zip(np.nditer(a), np.nditer(b)):
           ^
   
   During: typing of assignment at /Users/czgdp1807/numba_project/numba/numba/np/arraymath.py (932)
   
   File "numba/np/arraymath.py", line 932:
       def np_allclose_impl(a, b, rtol=1e-05, atol=1e-08, equal_nan=False):
           <source elided>
   
           for av, bv in zip(np.nditer(a), np.nditer(b)):
           ^

  raised from /Users/czgdp1807/numba_project/numba/numba/core/typeinfer.py:1086

During: resolving callee type: Function(<function allclose at 0x7f7e9072f560>)
During: typing of call at ../debug.py (6)


File "../debug.py", line 6:
def np_allclose(a, b, rtol=1e-05, atol=1e-08, equal_nan=False):
    return np.allclose(a, b, rtol, atol, equal_nan)
    ^
```

Reply:
I tried to workaround this problem in https://github.com/numba/numba/pull/7940/commits/f09d8ffd2555e59a70ec0618b1532e8d95511ea2 and it seems to work locally. `a_b`/`b_b` both have updates types (with correct ranks in the type), so `np_allclose_impl_no_broadcast` receives `a` and `b` parameters of the same shape and hence, `np.nditer((a, b))` works without issues.

Reply:
I used recently added `np.broadcast_arrays` functionality, so this conversation might not be relevant anymore.